bin/Macronify/Main.cpp

```diff
@@ -0,0 +1,734 @@
+        }
+
+        ast.emplace(maybe_ast.TakeValue());
+
+        mlir::DialectRegistry registry;
+
+        // Register the MLIR dialects we will be converting to
+        registry.insert<
+            vast::hl::HighLevelDialect,
+            macroni::macroni::MacroniDialect,
+            macroni::kernel::KernelDialect
+        >();
+        mctx.emplace(registry);
+        macroni::MacroniMetaGenerator meta(*ast, &*mctx);
+        vast::cg::CodeGenBase<macroni::Visitor> codegen(&*mctx, meta);
+        builder.emplace(&*mctx);
+
+        codegen.append_to_module(ast->UnderlyingAST().getTranslationUnitDecl());
+        mlir::OwningOpRef<mlir::ModuleOp> mod = codegen.freeze();
+
+        // TODO(bpp): Add a command-line argument to convert special macros into
+        // special operations
+
+        // Register conversions
+
+        // Mark expansions of get_user(), offsetof() and container_of() as
+        // illegal
+        mlir::ConversionTarget trg(*mctx);
+        trg.addDynamicallyLegalOp<macroni::macroni::MacroExpansion>(
+            [](Op *op) {
+                using ME = macroni::macroni::MacroExpansion;
+                ME exp = mlir::dyn_cast<ME>(op);
+                bool has_results = !exp.getResultTypes().empty();
+                llvm::StringRef macro_name = exp.getMacroName();
+                size_t num_params = exp.getParameterNames().size();
+                bool is_get_user = (has_results &&
+                                    num_params == 2 &&
+                                    macro_name == get_user),
+                    is_offsetof = (has_results &&
+                                   num_params == 2 &&
+                                   macro_name == offsetof),
+                    is_container_of = (has_results &&
+                                       num_params == 3 &&
+                                       macro_name == container_of),
+                    is_rcu_dereference = (has_results &&
+                                          num_params == 1 &&
+                                          macro_name == rcu_dereference),
+                    //   NOTE(bpp): The actual smp_mb() macro may not have
+                    //   results! This check will need to be updated if so.
+                    is_smp_mb = (has_results &&
+                                 num_params == 0 &&
+                                 macro_name == smp_mb);
+
+                using MP = macroni::macroni::MacroParameter;
+                using RMO = vast::hl::RecordMemberOp;
+                using PT = vast::hl::PointerType;
+                using TA = mlir::TypeAttr;
+                if (is_get_user) {
+                    exp.getExpansion().walk([&](Op *cur) {
+                        if (auto param_op = mlir::dyn_cast<MP>(cur)) {
+                            auto param_name = param_op.getParameterName();
+                            if (param_name == "x") {
+                                get_user_to_x.insert({ op, cur });
```

Comment:
What if we have multiple of strings we want to handle? `users_of = std::map< std::string, std::map< ... > >` and use that `"x"` and `"ptr"` as top-level keys. As a nice bonus you can then use a loop here instead of explicit branching on each possibility.